WPFSKillTree/Utils/UrlProcessing/BuildUrlNormalizer.cs

```diff
@@ -0,0 +1,125 @@
+        {
+            buildUrl = Regex.Replace(buildUrl, @"\s", "");
+
+            while (true)
+            {
+                if (buildUrl.Contains("google.com"))
+                {
+                    buildUrl = ExtractUrlFromQuery(buildUrl, "q");
+                    continue;
+                }
+
+                if (buildUrl.Contains("tinyurl.com") || buildUrl.Contains("poeurl.com") || buildUrl.Contains("goo.gl"))
+                {
+                    buildUrl = await ResolveShortenedUrl(buildUrl, loadingWrapper);
+                    continue;
+                }
+
+                break;
+            }
+
+            return EnsureProtocol(buildUrl);
+        }
+
+        protected virtual string ExtractUrlFromQuery(string url, string parameterName)
+        {
+            var match = Regex.Match(url, $@"{parameterName}=(?<urlParam>.*?)(&|$)");
+            if (!match.Success)
+                throw new ArgumentException($"The URL doesn't contain required query parameter '{parameterName}'."); // internal exception
+
+            return match.Groups["urlParam"].Value;
+        }
+
+        protected virtual async Task<string> ResolveShortenedUrl(string buildUrl, Func<string, Task, Task> loadingWrapper)
+        {
+            buildUrl = EnsureProtocol(buildUrl);
+
+            bool allowForbidden = false;
+
+            if (_poeUrlRegex.IsMatch(buildUrl))
+            {
+                // Fix for a bug in poeplanner: it returns 403 when getting 88+ points, but content is valid
+                allowForbidden = true;
+
+                buildUrl = buildUrl.Replace("preview.", "");
+                if (!buildUrl.Contains("redirect.php"))
+                {
+                    buildUrl = _poeUrlRegex.Replace(buildUrl, "http://www.poeurl.com/redirect.php?url=");
+                }
+            }
+
+            HttpResponseMessage response = null;
+
+            if (loadingWrapper != null)
+            {
+                // This lambda is used to skip GetAsync() result value in loadingWrapper parameter
+                // and thereby unify it by using Task instead of Task<HttpResponseMessage> in signature
+                Func<Task> headersLoader = async () =>
+                {
+                    response = await new HttpClient().GetAsync(buildUrl, HttpCompletionOption.ResponseHeadersRead);
+                };
+
+                await loadingWrapper(L10n.Message("Resolving shortened tree address"), headersLoader());
+
+                if (!allowForbidden)
+                {
+                    response.EnsureSuccessStatusCode();
+                }
+            }
+            else
+            {
+                response = await new HttpClient().GetAsync(buildUrl, HttpCompletionOption.ResponseHeadersRead);
+            }
+
+            return response.RequestMessage.RequestUri.ToString();
+        }
+
+        protected virtual string EnsureProtocol(string buildUrl)
```

Comment:
I've made a lot of methods `protected virtual` thinking first of all about extensibility and then of testability. Therefore, I've made tests for these methods just because it was possible and not vice versa.